aten/src/ATen/core/ivalue.h

```diff
@@ -699,6 +699,9 @@ struct TORCH_API IValue final {
   template <class T, size_t N>
   IValue(std::array<T, N> v);
 
+  // TODO: this is not robust
+  IValue(at::ArrayRef<c10::SymInt> v);
```

Comment:
For the array, `IValue(std::array<T, N> v)` will be called which always traverses all the elements and thus return a valid IValue (containing SymInt even if it was all int).

Can we just plan to keep specializing the container templates to properly double check the SymInt?

Reply:
Yeah, if we want to do it that way, I can add specializations for everything else.

Reply:
Oh, just to be clear, it is REQUIRED that we return an IValue that is an IntList if all of the elements happen to be integers only, because downstream code will attempt to toIntList if it is going straight to a kernel that doesn't support symints, and this will choke if you pass it a SymIntList

Reply:
> Yeah, if we want to do it that way, I can add specializations for everything else.

I'm not sure how much c++ magic can be done here. Are there other options than specializing everything?

Reply:
My other thought was to modify c10::List to know about how to look on the inside of the list elements to adjust the tag, but this has the efficiency problem I mentioned above.

Reply:
Well, the current code (and I guess the new specializations for other containers) are already quite slow?
You call is_symbolic() on every single element to check if they are symbolic in the `asIntArrayRefSlowOpt` call.

So if you just modify the `list.push_back(SymInt)` to check for is_symbolic (and update the tag accordingly), it is the same cost? We do traverse every element in any case.

Reply:
> Well, the current code (and I guess the new specializations for other containers) are already quite slow?
You call is_symbolic() on every single element to check if they are symbolic in the asIntArrayRefSlowOpt call.

This can be optimized though; we can add a boolean field to SymIntArrayRef to tag if everything is int, then we don't need to do a loopy check.

Reply:
I think my new plan is to static assert that T is not SymInt on the other containers and move on with my life.

Reply:
Note that I'm not saying that we must do either here. If the all symint PR can land, I think we should land this one as is! :)